# DS 3100 Activity Solution: Exploring the Tennessee Education Data

This notebook provides a solution to the data-frame activity using **Python/pandas** and the provided `tenn2018.csv` file.

> **Note:** Place `tenn2018.csv` in a data folder that is in the same directory as this notebook before running it.

## 1. Read the data and print the first few observations

We use `pd.read_csv()` to read the CSV file into a pandas **DataFrame** called `ten`. The `head()` method displays the first five observations by default.

In [2]:
import pandas as pd

ten = pd.read_csv("data/tenn2018.csv")
ten.head()

,district_number,district_name,school_number,school_name,year,subgroup,overall_subject,percent_below,percent_approaching,percent_on_track,percent_mastered,percent_on_mastered,baseline_year,percent_below_previous,percent_approaching_previous,percent_on_track_previous,percent_mastered_previous,percent_on_mastered_previous
0,0,State of Tennessee,0,NaN,2018,All Students,ELA,19.5,47.7,28.1,4.7,32.8,2017.0,20.0,45.9,28.0,6.1,34.1
1,0,State of Tennessee,0,NaN,2018,Black/Hispanic/Native,ELA,30.1,50.5,17.5,1.9,19.4,2017.0,30.9,49.2,17.6,2.3,19.9
2,0,State of Tennessee,0,NaN,2018,Economically Disadvantaged,ELA,30.4,51.2,16.7,1.7,18.4,2017.0,31.5,49.6,16.9,2.0,18.9
3,0,State of Tennessee,0,NaN,2018,English Learners,ELA,30.1,51.7,16.7,1.6,18.3,2017.0,34.3,49.1,15.1,1.5,16.6
4,0,State of Tennessee,0,NaN,2018,Foster,ELA,33.9,50.0,14.4,1.8,16.2,NaN,NaN,NaN,NaN,NaN,NaN


The first few rows show the structure of the dataset: each row is an observation, while each column is a variable.

## 2. Determine the number of rows and columns

The DataFrame attribute `.shape` returns `(number_of_rows, number_of_columns)`.

In [3]:
ten.shape

(121062, 18)

**Answer:** The dataset contains **121,062 rows** and **18 columns**.

## 3. Print the column names

`ten.columns` returns the names of all variables in the DataFrame. `len(ten.columns)` gives the number of variables.

In [12]:
print("Column names:", ten.columns.tolist())
print("Number of columns:", len(ten.columns))

Column names: ['district_number', 'district_name', 'school_number', 'school_name', 'year', 'subgroup', 'overall_subject', 'percent_below', 'percent_approaching', 'percent_on_track', 'percent_mastered', 'percent_on_mastered', 'baseline_year', 'percent_below_previous', 'percent_approaching_previous', 'percent_on_track_previous', 'percent_mastered_previous', 'percent_on_mastered_previous']
Number of columns: 18


**Answer:** There are **18 variables** in the dataset. Looking into the data, there appears to be 4 columns holding categorical values.

## 4. Access the `overall_subject` column

In pandas, square brackets can be used to select a column by name. `ten["overall_subject"]` returns the complete column as a pandas Series.

In [5]:
ten["overall_subject"]

0                    ELA
1                    ELA
2                    ELA
3                    ELA
4                    ELA
               ...      
121057    Social Studies
121058    Social Studies
121059    Social Studies
121060    Social Studies
121061    Social Studies
Name: overall_subject, Length: 121062, dtype: str

## 5. Access `district_name` for the first observation

A pandas DataFrame can be indexed by row and column. Using `.loc` lets us specify the row and column labels explicitly. With the default integer index, the first observation has index `0`.

In [6]:
ten.loc[0, "district_name"]

'State of Tennessee'

**Answer:** `"State of Tennessee"`

## 6. Access `percent_below` for the first observation

In [7]:
ten.loc[0, "percent_below"]

np.float64(19.5)

**Answer:** `19.5`

## 7. How many columns contain missing values?

`ten.isna()` creates a Boolean DataFrame showing where values are missing.

`.any()` checks whether each column contains at least one missing value, and `.sum()` counts how many columns satisfy that condition.

In [10]:
missing_by_column = ten.isna().any(axis=0).sum()
print("Number of columns with missing values:", missing_by_column)

Number of columns with missing values: 15


**Answer:** **15 columns** contain at least one missing value.

The other **3 columns**—`district_number`, `school_number`, and `year`—have no missing values.